File: app/schemas/car.py
from pydantic import BaseModel
from typing import Optional
from datetime import datetime

class CarBrandBase(BaseModel):
    name: str
    normalized_name: str

class CarBrandCreate(CarBrandBase):
    pass

class CarBrand(CarBrandBase):
    id: int

    class Config:
        from_attributes = True

class CarModelBase(BaseModel):
    name: str
    normalized_name: str
    brand_id: int

class CarModelCreate(CarModelBase):
    pass

class CarModel(CarModelBase):
    id: int

    class Config:
        from_attributes = True

class CarListingBase(BaseModel):
    yad2_id: str
    title: str
    description: Optional[str] = None
    price: float
    year: int
    mileage: Optional[int] = None
    fuel_type: Optional[str] = None
    transmission: Optional[str] = None
    body_type: Optional[str] = None
    color: Optional[str] = None
    image_url: Optional[str] = None  # URL to the main car image
    status: str = "active"
    brand_id: int
    model_id: int
    last_scraped_at: Optional[datetime] = None

class CarListingCreate(CarListingBase):
    pass

class CarListing(CarListingBase):
    id: int
    created_at: datetime
    updated_at: Optional[datetime] = None
    brand: CarBrand
    model: CarModel

    class Config:
        from_attributes = True
        
        @classmethod
        def model_validate(cls, obj, *, strict=None, from_attributes=None, context=None):
            # Ensure updated_at is set to created_at if not provided
            if hasattr(obj, 'created_at') and (not hasattr(obj, 'updated_at') or obj.updated_at is None):
                obj.updated_at = obj.created_at
            return super().model_validate(obj, strict=strict, from_attributes=from_attributes, context=context)

File: frontend/src/types/car.ts
export interface CarBrand {
  id: number;
  name: string;
  normalized_name: string;
}

export interface CarModel {
  id: number;
  brand_id: number;
  name: string;
  normalized_name: string;
}

export interface CarListing {
  id: number;
  yad2_id: string;
  brand_id: number;
  model_id: number;
  title: string;
  description: string;
  price: number;
  original_price?: number;
  is_price_negotiable: boolean;
  year: number;
  mileage: number;
  fuel_type?: string;
  transmission?: string;
  body_type?: string;
  color?: string;
  engine_volume?: number;
  horsepower?: number;
  doors?: number;
  seats?: number;
  is_imported: boolean;
  is_accident_free: boolean;
  test_until?: string;
  city?: string;
  neighborhood?: string;
  location: string;
  thumbnail_url?: string;
  source_url: string;
  source_created_at?: string;
  source_updated_at?: string;
  status: 'active' | 'sold' | 'pending' | 'inactive' | 'expired';
  created_at: string;
  updated_at: string;
  last_seen_at: string;
  metadata?: Record<string, any>;
  // Relations
  brand_name?: string;
  model_name?: string;
}

export interface CarListingFilters {
  brand?: string;
  model?: string;
  minPrice?: number;
  maxPrice?: number;
  yearFrom?: number;
  yearTo?: number;
  location?: string;
  page?: number;
  limit?: number;
}
